src/engine.rs

```diff
@@ -343,15 +346,85 @@ where
     }
 
     /// Truncates Raft groups to remove possible corruptions.
-    #[allow(unused_variables)]
     pub fn unsafe_truncate_with_file_builder(
         path: &Path,
-        mode: &str,
+        mode: TruncateMode,
         queue: Option<LogQueue>,
         raft_groups: &[u64],
         file_builder: Arc<B>,
     ) -> Result<()> {
-        todo!();
+        if !path.exists() {
+            return Err(Error::InvalidArgument(format!(
+                "raft-engine directory '{}' does not exist.",
+                path.to_str().unwrap()
+            )));
+        }
+
+        let cfg = Config {
+            dir: path.to_str().unwrap().to_owned(),
+            recovery_mode: RecoveryMode::TolerateAnyCorruption,
+            ..Default::default()
+        };
+
+        let parms = TruncateQueueParameter {
+            queue,
+            truncate_mode: mode,
+            raft_groups_ids: raft_groups.to_vec(),
+            lock_file: Arc::new(get_lock_file(path.to_str().unwrap())?),
+        };
+
+        let mut builder = FilePipeLogBuilder::new(cfg.clone(), file_builder.clone(), Vec::new());
+        builder.scan()?;
+
+        let threads = cfg.recovery_threads;
```

Comment:
I'm very confused about this part, why can't we reuse `recover_queue`?

Reply:
Okay, you can make it cleaner by removing the thread pool building code. We don't need all that parallelism for a debugging task like "truncate". Instead, you can call two `recover_queue` sequentially (and preferably passing in concurrency=1).